src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs

```diff
@@ -1912,35 +1973,68 @@ public static Matrix4x4 Subtract(Matrix4x4 value1, Matrix4x4 value2)
         /// <param name="value1">The first source matrix.</param>
         /// <param name="value2">The second source matrix.</param>
         /// <returns>The result of the multiplication.</returns>
-        public static Matrix4x4 Multiply(Matrix4x4 value1, Matrix4x4 value2)
+        public static unsafe Matrix4x4 Multiply(Matrix4x4 value1, Matrix4x4 value2)
         {
-            Matrix4x4 result;
-
-            // First row
-            result.M11 = value1.M11 * value2.M11 + value1.M12 * value2.M21 + value1.M13 * value2.M31 + value1.M14 * value2.M41;
-            result.M12 = value1.M11 * value2.M12 + value1.M12 * value2.M22 + value1.M13 * value2.M32 + value1.M14 * value2.M42;
-            result.M13 = value1.M11 * value2.M13 + value1.M12 * value2.M23 + value1.M13 * value2.M33 + value1.M14 * value2.M43;
-            result.M14 = value1.M11 * value2.M14 + value1.M12 * value2.M24 + value1.M13 * value2.M34 + value1.M14 * value2.M44;
-
-            // Second row
-            result.M21 = value1.M21 * value2.M11 + value1.M22 * value2.M21 + value1.M23 * value2.M31 + value1.M24 * value2.M41;
-            result.M22 = value1.M21 * value2.M12 + value1.M22 * value2.M22 + value1.M23 * value2.M32 + value1.M24 * value2.M42;
-            result.M23 = value1.M21 * value2.M13 + value1.M22 * value2.M23 + value1.M23 * value2.M33 + value1.M24 * value2.M43;
-            result.M24 = value1.M21 * value2.M14 + value1.M22 * value2.M24 + value1.M23 * value2.M34 + value1.M24 * value2.M44;
-
-            // Third row
-            result.M31 = value1.M31 * value2.M11 + value1.M32 * value2.M21 + value1.M33 * value2.M31 + value1.M34 * value2.M41;
-            result.M32 = value1.M31 * value2.M12 + value1.M32 * value2.M22 + value1.M33 * value2.M32 + value1.M34 * value2.M42;
-            result.M33 = value1.M31 * value2.M13 + value1.M32 * value2.M23 + value1.M33 * value2.M33 + value1.M34 * value2.M43;
-            result.M34 = value1.M31 * value2.M14 + value1.M32 * value2.M24 + value1.M33 * value2.M34 + value1.M34 * value2.M44;
-
-            // Fourth row
-            result.M41 = value1.M41 * value2.M11 + value1.M42 * value2.M21 + value1.M43 * value2.M31 + value1.M44 * value2.M41;
-            result.M42 = value1.M41 * value2.M12 + value1.M42 * value2.M22 + value1.M43 * value2.M32 + value1.M44 * value2.M42;
-            result.M43 = value1.M41 * value2.M13 + value1.M42 * value2.M23 + value1.M43 * value2.M33 + value1.M44 * value2.M43;
-            result.M44 = value1.M41 * value2.M14 + value1.M42 * value2.M24 + value1.M43 * value2.M34 + value1.M44 * value2.M44;
-
-            return result;
+            if (Sse.IsSupported)
+            {
+                Matrix4x4 m = default;
+
+                Sse.Store(&m.M11,
+                    Sse.Add(Sse.Add(Sse.Multiply(Sse.SetAllVector128(value1.M11), Sse.LoadVector128(&value2.M11)),
```

Comment:
Rather than calling `SetAllVector128` multiple times, it is probably better to do a load row and permute (broadcast in AVX, shuffle in SSE)

Reply:
btw, it'd be nice to expose `_MM_SHUFFLE` macros to C# for Shuffle.
e.g. `static byte _MM_SHUFFLE(byte fp3, byte fp2, byte fp1, byte fp0) => (byte)(((fp3) << 6) | ((fp2) << 4) | ((fp1) << 2) | ((fp0)));`

Reply:
I've raised it before. It should go to API review soon

Reply:
Agreed. @tannergooding and I had the same conclusion during https://github.com/dotnet/machinelearning/pull/562.  However, I didn't see an issue open for this in either corefx or coreclr.  @tannergooding did you ever open one?

Here is what we proposed at the time:
```
 `Sse.Shuffle(op1, op2, Sse.GetShuffleControl(0, 3, 1, 2))` 
Something like that. `GetShuffleControl`, `BuildShuffleControl`, or something
```

Reply:
So I've tried to use shuffle here but it became slower, maybe I did something wrong? 
![image](https://user-images.githubusercontent.com/523221/44172305-707d1280-a0e5-11e8-915c-acb2f8498bc7.png)

Reply:
I'll take a look at the assembly and see if I can spot what is going wrong....

Reply:
Logged a bug to track this here: https://github.com/dotnet/corefx/issues/31825

Reply:
@tannergooding any update on `_MM_SHUFFLE ` API? Could not find an issue for it